controllers/backupstorage_controller.go

```diff
@@ -97,35 +103,78 @@ func (r *BackupStorageReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 		}
 		return ctrl.Result{}, err
 	}
-	var needsUpdate bool
-	if req.NamespacedName.Namespace == r.systemNamespace {
-		logger.Info("setting controller references for the secret")
-		needsUpdate, err = r.reconcileBackupStorage(ctx, bs, defaultSecret)
-		if err != nil {
+
+	if req.NamespacedName.Namespace != r.systemNamespace {
+		return ctrl.Result{}, nil
+	}
+
+	// Set controllerRef on the defaultSecret.
+	if !metav1.IsControlledBy(defaultSecret, bs) {
+		logger.Info("setting controller reference for the secret")
+		if err := controllerutil.SetControllerReference(bs, defaultSecret, r.Scheme); err != nil {
 			return ctrl.Result{}, err
 		}
-		if !controllerutil.ContainsFinalizer(bs, cleanupSecretsFinalizer) {
-			controllerutil.AddFinalizer(bs, cleanupSecretsFinalizer)
-			if err := r.Update(ctx, bs); err != nil {
-				return ctrl.Result{}, err
-			}
+		if err := r.Update(ctx, defaultSecret); err != nil {
+			return ctrl.Result{}, err
 		}
 	}
 
-	logger.Info("updating secrets in used namespaces")
-	if err := r.handleSecretUpdate(ctx, bs, defaultSecret); err != nil {
+	// Add finalizer.
+	if controllerutil.AddFinalizer(bs, cleanupSecretsFinalizer) {
+		if err := r.Update(ctx, bs); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	needStatusUpdate, err := r.reconcileUsedNamespaces(ctx, bs)
+	if err != nil {
 		return ctrl.Result{}, err
 	}
-	if needsUpdate {
-		err := r.Status().Update(ctx, bs)
-		if err != nil {
+	if bs.UpdateNamespacesList(defaultSecret.Namespace) || needStatusUpdate {
+		if err := r.Status().Update(ctx, bs); err != nil {
 			return ctrl.Result{}, err
 		}
 	}
 
+	logger.Info("updating secrets in used namespaces")
+	if err := r.handleSecretUpdate(ctx, bs, defaultSecret); err != nil {
+		return ctrl.Result{}, err
+	}
+
```

Comment:
What's the reason for doing this after the status update?

Reply:
`handleSecretUpdate` basically updates all the secrets in the `UsedNamespaces`. In order to do that, you need to refresh the list of UsedNamespaces first, which is why this step happens in the end.

For example, if a namespace has been deleted, and UsedNamespace is not updated first, then `handleSecretUpdate` will try to reconcile a non-existent Secret..

Reply:
🤔 I see a possible issue though. Let's say you add namespace A. You first update the status and then you try to create the secret. What if the secret creation fails, e.g. by lack of permissions? You are reporting in the status that namespace A has the secret but actually it does not...
I don't think it's correct to update the status before updating the secret, it can lead to race conditions...

Reply:
But we don't create Secrets here. The secrets are created/copied in their respective DB controllers.

What's really happening here is:
- List all namespaces where the secret exists already (these is the `UsedNamespace`)
- Update the status
- Then inside `handleSecretUpdate`: for each used Namespace, update the secret.

Only if the Secret has already been created (which doesn't happen here), will it be a part of the `UsedNamespace`, and then be reconciled (only updated) later. If the Secret creation fails, then it won't even be a part of the status in the first place.

Does this make sense? Let me know if I'm still missing something..

Reply:
TBH, I don't know why we need the `handleSecretUpdate` step here, since they are already being reconciled inside the DB controller. Maybe best to remove it from here to avoid confusion..

Reply:
I think you're right, I'll double-check it nonetheless.
In the meantime maybe you know the answer to this one as well. What happens if someone manually changes the secret in a given namespace? Will the reconciliation be triggered on both controllers? Will the drift be automatically reconciled? By which controller?

Reply:
Yeah I think this analysis is correct, so this PR can be merged as is, no issues here.